kyo-core/shared/src/main/scala/kyo/Meter.scala

```diff
@@ -115,25 +108,23 @@ object Meter:
       * @return
       *   A Meter effect that represents a rate limiter.
       */
-    def initRateLimiter(rate: Int, period: Duration)(using Frame): Meter < IO =
-        Channel.init[Unit](rate).map { chan =>
-            Timer.scheduleAtFixedRate(period)(offer(rate, chan, ())).map { _ =>
-                new Meter:
-
-                    def available(using Frame)              = chan.size
-                    def run[A, S](v: => A < S)(using Frame) = chan.take.map(_ => v)
-
-                    def tryRun[A, S](v: => A < S)(using Frame) =
-                        chan.poll.map {
-                            case Absent =>
-                                Maybe.empty
-                            case _ =>
-                                v.map(Maybe(_))
-                        }
-
-                    def close(using Frame) =
-                        chan.close.map(_.isDefined)
-            }
+    def initRateLimiter(rate: Int, period: Duration)(using initFrame: Frame): Meter < IO =
+        IO.Unsafe {
+            new Base(rate):
+                val timerTask =
+                    // Schedule periodic task to replenish permits
+                    Timer.live.unsafe.scheduleAtFixedRate(period, period)(replenish())
+
+                def dispatch[A, S](v: => A < S) =
+                    // Don't release a permit since it's managed by the timer task
+                    v
+
+                @tailrec def replenish(i: Int = 0): Unit =
+                    if i < rate then
+                        release()
+                        replenish(i + 1)
+
```

Comment:
I could be wrong, but I suspect this may lead to over-replenishing if the permits are not fully used in a given period.

Reply:
Good catch! Fixed 🙏